### ISSUE TITLE
Rails-style validations for inputs

### ISSUE THREAD

**Issue by nvh0412** [rmosolgo/graphql-ruby#3164]
**Is your feature request related to a problem? Please describe.**

Based on this issue https://github.com/rmosolgo/graphql-ruby/issues/799, I just would like to implement this stuff for our repo. But the issue should be created beforehand to make sure that I don't do anything redundant or wrong things for this feature.

**Describe the solution you'd like**

Currently, our `argument` DSL allows us to pass a block into it, so I want to leverage it and make a deeper DSL level for our input validation. Let's say we want to validate input of 1 argument for bills field like below:

```ruby
field :bills, [BillType], 'Returns all bills', null: false do
  argument :phoneNumber, String, required: false
end
```

Then we can add a new validation for `phoneNumber` argument
```ruby
field :bills, [BillType], 'Returns all bills', null: false do
  argument :phoneNumber, String, required: false do
    validate format: { with: /abc/,  message: "xyz" }
  end
end
```

The validation keyword/syntax is getting inspire by ActiveRecord validations (to help us on remembering our syntax easier)

As for how to validate value of input before it's resolved, I have a plan to use our Field Extension mechanism, we're gonna have a build-in field extension i.e ArgumentValidationExtension, and add it into @extension array of its parent field by default. By this way, we can leverage extension to validation and raise exception if the input is invalid

**Describe alternatives you've considered**

I considered about using a hash which will include our validation stuff and pass it as a params to our `argument` DSL before our &block, but it would

**Comment by rmosolgo** [comment]
Hey, thanks for the suggestion. I mentioned the same thing in the 1.12 release issue (#3056), and I totally agree with your approach: 

- Use as much syntax/features from ActiveRecord as possible (is it possible to even use ActiveRecord/ActiveModel's validator classes?) 
- A field extension seems like a great approach 

Are you interested in working on this? I'd be happy to review a PR if you want to give it a shot!

**Comment by nvh0412** [comment]
@rmosolgo thanks. Of course that I want to give it a shot and I finished the basic structure for it. But it would take more time than I expect, because I'm getting married haha. So when I create PR for this validation stuff, please help me review it. By the way, thanks for creating a great repo. I learnt a lot from codebase.

**Comment by rmosolgo** [comment]
Addressed by #3207, will be released in 1.12.0
